internal/driver/driver.go

```diff
@@ -0,0 +1,646 @@
+				d.lc.Errorf("failed to get device serial number, error: %s", err.Error())
+				continue
+			}
+			if _, ok := cdm[cn+sn]; ok {
+				continue
+			}
+			discovered := sdkModels.DiscoveredDevice{
+				Name: buildDeviceName(cn, sn),
+				Protocols: map[string]models.ProtocolProperties{
+					UsbProtocol: {
+						Path:         fdPath,
+						SerialNumber: sn,
+						CardName:     cn,
+					},
+				},
+				Description: fmt.Sprintf("USB camera %s", cn),
+				Labels:      []string{"auto-discovery", cn},
+			}
+			devices = append(devices, discovered)
+			d.lc.Infof("discovered device: %s", discovered.Name)
+		}
+	}
+	d.deviceCh <- devices
+}
+
+func (d *Driver) getProtocolProperty(protocols map[string]models.ProtocolProperties, protocol, key string) (string, errors.EdgeX) {
+	if _, ok := protocols[protocol]; !ok {
+		return "", errors.NewCommonEdgeX(errors.KindContractInvalid,
+			fmt.Sprintf("%s protocol configuration not found. Please check device configuration", protocol), nil)
+	}
+	value, ok := protocols[protocol][key]
+	if !ok {
+		return "", errors.NewCommonEdgeX(errors.KindContractInvalid,
+			fmt.Sprintf("property %s of protocol %s is missing. Please check device configuration",
+				key, protocol), nil)
+	}
+	return value, nil
+}
+
+func (d *Driver) newDevice(name string, protocols map[string]models.ProtocolProperties) (*Device, errors.EdgeX) {
+	fdPath, edgexErr := d.getProtocolProperty(protocols, UsbProtocol, Path)
+	if edgexErr != nil {
+		return nil, errors.NewCommonEdgeXWrapper(edgexErr)
+	}
+
+	rtspUri := &url.URL{
+		Scheme: RtspUriScheme,
+		Host:  fmt.Sprintf("%s:%s",d.rtspHostName,d.rtspTcpPort),
+	}
+	rtspUri.Path = path.Join(Stream, name)
+
+	// Create new instance of transcoder
+	trans := new(transcoder.Transcoder)
+	// Initialize transcoder passing the input path and output path
+	err := trans.Initialize(fdPath, rtspUri.String())
+	if err != nil {
+		return nil, errors.NewCommonEdgeX(errors.KindServerError,
+			fmt.Sprintf("failed to initialize transcoder for device %s", name), err)
+	}
+	trans.MediaFile().SetOutputFormat(RtspUriScheme)
+
+	autoStreaming := false
+	autoStreamingStr, edgexErr := d.getProtocolProperty(protocols, UsbProtocol, AutoStreaming)
+	if edgexErr != nil {
+		d.lc.Warnf("Protocol property %s not found. Use default value: %v", AutoStreaming, autoStreaming)
+	} else {
+		autoStreaming, err = strconv.ParseBool(autoStreamingStr)
+		if err != nil {
+			d.lc.Errorf("invalid input value %s for protocol property %s. Use default value: %v", autoStreamingStr,
+				AutoStreaming, autoStreaming)
+		}
+	}
+
+	device, err := d.ds.GetDeviceByName(name)
+	if err != nil {
+		return nil, errors.NewCommonEdgeX(errors.KindServerError,
+			fmt.Sprintf("device %s not found in core metadata", name), err)
+	}
+	profile, err := d.ds.GetProfileByName(device.ProfileName)
+	if err != nil {
+		return nil, errors.NewCommonEdgeX(errors.KindServerError,
+			fmt.Sprintf("profile %s not found in core metadata", name), err)
+	}
```

Comment:
Shouldn't this be done first thing?

Reply:
You are right. Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/8414d18ef7d553f530fd641a84d2cf3a0311ec64